### PR TITLE
LNX-132 Make ticking depend on get. Add CreateObject action when adding object

### PR DESCRIPTION
**Pull request by alpinus4**
It is a necessary change for the multiplayer to work.
The frontend has to have disabled ticking functionality, and handle CreateObject actions.

### CODE REVIEW

**Review by filip-szweda — APPROVED** [review]
Looks good to me

**Review by EasyJB — APPROVED** [review]
For now looking good.

**Comment by alpinus4** [comment]
I removed adding object with tick parameter from spam objects test, as the test just wasn't working with fastapi backgroundTasks.

File: src/app.py [hunk 1]
```diff
@@ -32,6 +34,7 @@ class GlobalState:
     processes: Dict = None,
     transitions: Dict = None,
     tick_number: int = 0
+    last_tick: float = 0
```
Comment (filip-szweda) [inline src/app.py]: `last_tick: float = 0.0`

File: src/app.py [hunk 2]
```diff
@@ -17,6 +18,7 @@
 from lynx.common.object import Object
 from lynx.common.scene import Scene
 from lynx.common.vector import Vector
+from lynx.common.actions.create_object import CreateObject
 from pydantic import BaseModel
```
Comment (filip-szweda) [inline src/app.py]: Separate standard library, imported library and our own library imports if you have a chance

File: src/app.py [hunk 3]
```diff
@@ -100,17 +103,43 @@ async def send_scene():
 
     await asyncio.gather(*future_sends)
 
+
```
Comment (filip-szweda) [inline src/app.py]: I think we should commit reformatting changes separately, this shouldn't be here

File: src/app.py [hunk 4]
```diff
@@ -100,17 +103,43 @@ async def send_scene():
...
+async def tick_trigger():
+    if time.time() - state.last_tick >= 1:
+        await tick()
+        state.last_tick = time.time()
+
+async def wait_for_next_tick():
+    tick = state.tick_number
+    while state.tick_number != tick + 1:
```
Comment (filip-szweda) [inline src/app.py]:
```
next_tick_number = state.tick_number + 1
while state.tick_number != next_tick_number:
```

File: src/app.py [hunk 5]
```diff
@@ -100,17 +103,43 @@ async def send_scene():
...
+        await asyncio.sleep(0.1)
```
Comment (filip-szweda) [inline src/app.py:129]: Why `0.1` specifically?

File: src/app.py [hunk 6]
```diff
@@ -100,17 +103,43 @@ async def send_scene():
...
+
+async def spawn_process_for_new_agent(object: Object):
+    await wait_for_next_tick()
+
+    parent_conn, child_conn = AioPipe()
```
Comment (filip-szweda) [inline src/app.py]: Please expand on what `conn` is

File: src/app.py [hunk 7]
```diff
@@ -100,17 +103,43 @@ async def send_scene():
...
+    p = AioProcess(target=execution_runtime,
```
Comment (filip-szweda) [inline src/app.py]: `process`, we're cool with longer variable names

File: src/app.py [hunk 8]
```diff
@@ -100,17 +103,43 @@ async def send_scene():
...
+                    args=(child_conn, object.id,))
```
Comment (filip-szweda) [inline src/app.py]: Should be inline with `p = AioProcess(target=execution_runtime,`

File: src/app.py [hunk 9]
```diff
@@ -100,17 +103,43 @@ async def send_scene():
...
+    p.start()
+
+    await parent_conn.coro_send(state.scene.serialize())
+    state.processes[object.id] = ProcessData(
+        process=p, pipe=parent_conn)
```
Comment (filip-szweda) [inline src/app.py]: Should be inline with `state.processes[object.id] = ProcessData(`
I think you should set longer line breaks in IDE

**Review by filip-szweda — CHANGES_REQUESTED** [review]

File: tests/scene_server_test.py [hunk 10]
```diff
@@ -42,14 +42,11 @@ async def test_spam_objects(self):
...
-
                 await TestSceneServer.post(ac, "/populate")
-                agent = Object(id=1000, tick=f"move(Vector(1,0))")
-                await TestSceneServer.post(ac, "/add_object", {'serialized_object': agent.serialize()})
-
                 response = await TestSceneServer.fetch(ac, "/?tick_number=-1")
                 response = await TestSceneServer.fetch(ac, "/?tick_number=0")
-                # scene = Scene.deserialize(response['scene'])
+
+                # we cannot test add_object with tick here, AsyncManager doesn't handle BackgroundTasks
```
Comment (blazej-smorawski) [inline tests/scene_server_test.py:49]: I wouldn't say it solves the problem, I think that we need such test. We could add some smaller tests if this one is not possible too.

File: src/app.py (hunk 5 above)
Comment (alpinus4) [reply src/app.py:129]: Just a small time amount, no particular reason

File: src/app.py (hunk 6 above)
Comment (alpinus4) [reply src/app.py]: > `parent_conn` and `child_conn` refer to the parent and child connections, respectively, in a multiprocessing setup. These connections are typically used for interprocess communication (IPC) between a parent process and a child process.

> In the code snippet you provided, `AioPipe()` is used to create a pair of asynchronous pipes for communication between the parent and child processes. The `AioPipe()` function likely creates two connected `asyncio` streams, one for the parent process (`parent_conn`) and one for the child process (`child_conn`).

> With these pipes, you can send messages or data between the parent and child processes asynchronously. The parent process can write data to the `parent_conn` pipe, and the child process can read from it using the `child_conn` pipe, and vice versa.

> This setup is commonly used in multiprocessing scenarios where you need to distribute tasks among multiple processes and exchange data or messages between them.

File: pytest.ini [hunk 11]
```diff
@@ -1,2 +1,4 @@
 [pytest]
-pythonpath = .
\ No newline at end of file
+pythonpath = .
```
Comment (filip-szweda) [inline pytest.ini:2]: Is it safe to use "."?

Comment (filip-szweda) [inline pytest.ini:2]: Why spaces before and after "="?

File: src/app.py [hunk 12]
```diff
@@ -33,6 +36,7 @@ class GlobalState:
     processes: Dict = None,
     transitions: Dict = None,
     tick_number: int = 0
+    last_tick: float = 0.0
```
Comment (filip-szweda) [inline src/app.py]: `last_tick_time`

File: src/app.py [hunk 13]
```diff
@@ -109,18 +113,41 @@ async def send_scene():
...
+async def wait_for_next_tick():
+    next_tick_number = state.tick_number + 1
+    while state.tick_number != next_tick_number:
+        await asyncio.sleep(0.1)
+
+async def spawn_process_for_new_agent(object: Object):
+    await wait_for_next_tick()
+
+    parent_connection, child_connection = AioPipe()
+    process = AioProcess(target=execution_runtime,
```
Comment (filip-szweda) [inline src/app.py]: Why is the code broken between two lines?

File: src/app.py [hunk 14]
```diff
@@ -109,18 +113,41 @@ async def send_scene():
...
+    if time.time() - state.last_tick >= 1:
+        await tick()
+        state.last_tick = time.time()
+
+async def wait_for_next_tick():
+    next_tick_number = state.tick_number + 1
+    while state.tick_number != next_tick_number:
+        await asyncio.sleep(0.1)
+
+async def spawn_process_for_new_agent(object: Object):
```
Comment (filip-szweda) [inline src/app.py:131]: `-> NoReturn`

File: src/app.py [hunk 15]
```diff
@@ -130,20 +157,11 @@ async def get(tick_number: int) -> Dict[str, Union[int, str, List[List[str]]]]:
 
 
 @app.post("/add_object")
-async def add(r: AddObjectRequest):
+async def add(r: AddObjectRequest, background_tasks: BackgroundTasks):
```
Comment (filip-szweda) [inline src/app.py:159]: `-> NoReturn`

File: src/app.py [hunk 16]
```diff
@@ -130,20 +157,11 @@ async def get(tick_number: int) -> Dict[str, Union[int, str, List[List[str]]]]:
...
     object = Object.deserialize(r.serialized_object)
-    state.scene.add_entity(object)
+    state.scene.add_to_pending_actions(CreateObject(r.serialized_object).serialize())
     if object.tick != "":
```
Comment (filip-szweda) [inline src/app.py:162]: Would be cool to have a comment here explaining that we only spawn processes for agents with code
Also, we really need to change the name of this attribute (`tick`, I mean) in the future xD

**Review by filip-szweda — CHANGES_REQUESTED** [review]
The logic itself looks good

File: pytest.ini (hunk 11 above)
Comment (alpinus4) [reply pytest.ini:2]: I made my additions consistent with what was there before

File: src/app.py (hunk 14 above)
Comment (alpinus4) [reply src/app.py:131]: Let's add `NoReturn` to functions in separate refactor PR

File: src/app.py (hunk 15 above)
Comment (alpinus4) [reply src/app.py:159]: As before

File: pytest.ini [hunk 17]
```diff
@@ -1,4 +1,4 @@
 [pytest]
 pythonpath = .
-log_cli=true  
-log_level=INFO
\ No newline at end of file
+log_cli = true
```
Comment (filip-szweda) [inline pytest.ini:3]: Thought that "=" is better than " = "? A whatever, but made me thinking

**Review by filip-szweda — APPROVED** [review]
'Tis enough

File: pytest.ini (hunk 17 above)
Comment (alpinus4) [reply pytest.ini:3]: I like it more with spaces, it just looks better.
It depends, in some places it matters, in others it doesn't.
If I can choose I would go for a variant with spaces any day.

**Pull request merged**